File: Difficulty: Medium/Next Greater Element in Circular Array/next-greater-element-in-circular-array.cpp
class Solution {
  public:
    vector<int> nextLargerElement(vector<int> &arr) {
        int n = arr.size();
        vector<int> ans(n, -1);
        stack<int> st;
        
        for(int i=2*n-1; i>=0; i--) {
            int idx = i % n;
            while(!st.empty() && st.top() <= arr[idx]) {
                st.pop();
            }
            if(i < n && !st.empty()) {
                ans[i] = st.top();
            }
            st.push(arr[idx]);
        }
        
        return ans;
    }
};

File: Difficulty: Medium/Next Greater Element in Circular Array/next-greater-element-in-circular-array.java
class Solution {
    public ArrayList<Integer> nextGreater(int[] arr) {
        int n = arr.length;
        Stack<Integer> st = new Stack<>();
        ArrayList<Integer> ans = new ArrayList<>(Collections.nCopies(n, -1));
        
        for(int i=2*n-1; i>=0; i--) {
            int index = i % n;
            while(!st.isEmpty() && st.peek() <= arr[index]) st.pop();
            if(!st.isEmpty()) ans.set(index, st.peek());
            st.push(arr[index]);
        }
        
        return ans;
    }
}